xdsl/ir.py

```diff
@@ -15,12 +15,31 @@
 OpT = TypeVar('OpT', bound='Operation')
 
 
+@dataclass
+class Dialect:
+    """Contains the operations and attributes of a specific dialect"""
+    _operations: list[Operation] = field(default_factory=list,
```

Comment:
Just wondering, why make this private?

Reply:
AFAIK, one `_` is protected, right?

This was mainly an intuitive decision. I feel that having them private would not work, as we need to change them in the subclasses. Having them public would give too much freedom and lead to a non-intuitive interface as one could do the following:

```
dialect1 = Dialect1()
ctx.register_dialect(dialect1)
dialect1.operations.append(SomeOp)
```

Intuitively, I would expect that the context would then know about SomeOp and be able to parse it, which it does not in the current design. Therefore, I decided to go with protected.

I think we could enable use like this example by adding a layer of indirection (the context has dialects instead of operations/attributes and has to look up in them when parsing), but I did not want to do changes as deep as that atm.

Reply:
Actually, thinking more about it, it might make sense to override __init__ with a constructor that takes two lists and assigns them properly. Then, we could do the same in its subclass, calling hte parent constructor from there...

Reply:
I would argue that for now, it is actually useless to have a `Dialect` subclass for each dialect, since instantiating them is kind of pointless.
Since as of right now a dialect is just a collection of ops/attrs, we could just have a constructor taking a list of operations and attributes, and we would instantiate the dialect objects in their files (instead of having dialect classes).
That would mean we would write `ctx.register_dialect(func)` instead of `ctx.register_dialect(Func())`, which I currently find a bit ugly.

Reply:
Oh also, we could make `operation` a property returning an Operation iterator, so we can easily access it without modifying it.

Reply:
> I would argue that for now, it is actually useless to have a `Dialect` subclass for each dialect, since instantiating them is kind of pointless. 

Yes, that makes sense. If we actually find a case where dialect objects are useful, we can still refactor to have them.

Reply:
> Oh also, we could make `operation` a property returning an Operation iterator, so we can easily access it without modifying it.

This actually does not work in our current testing infrastructure. In both `rewriter_test` and `mlir_converter_test` we register the dialect multiple times. This means that the second time it is registered, the iterator is already at the end of the list, so nothing is registered and the parsing of the programs fails. I guess we can a) rewrite the tests or b) leave it as lists. I would argue for b) as it feels like a very weird limitation for users and I don't think we gain a lot from it.